File: 97. Interleaving String/97. Interleaving String.c
bool isInterleave(char * s1, char * s2, char * s3) {
    int len1 = strlen(s1), len2 = strlen(s2), len3 = strlen(s3);
    if (len1 + len2 != len3) return false;
    
    bool dp[len2 + 1];
    memset(dp, 0, sizeof(dp));
    
    for (int i = 0; i <= len1; i++) {
        for (int j = 0; j <= len2; j++) {
            if (i == 0 && j == 0) {
                dp[j] = true;
            } else if (i == 0) {
                dp[j] = dp[j-1] && s2[j-1] == s3[j-1];
            } else if (j == 0) {
                dp[j] = dp[j] && s1[i-1] == s3[i-1];
            } else {
                dp[j] = (dp[j] && s1[i-1] == s3[i+j-1]) || 
                        (dp[j-1] && s2[j-1] == s3[i+j-1]);
            }
        }
    }
    
    return dp[len2];
}

File: 97. Interleaving String/97. Interleaving String.ts
function isInterleave(s1: string, s2: string, s3: string): boolean {
    if (s1.length + s2.length !== s3.length) return false;
    
    const dp: boolean[] = new Array(s2.length + 1).fill(false);
    
    for (let i = 0; i <= s1.length; i++) {
        for (let j = 0; j <= s2.length; j++) {
            if (i === 0 && j === 0) {
                dp[j] = true;
            } else if (i === 0) {
                dp[j] = dp[j-1] && s2[j-1] === s3[j-1];
            } else if (j === 0) {
                dp[j] = dp[j] && s1[i-1] === s3[i-1];
            } else {
                dp[j] = (dp[j] && s1[i-1] === s3[i+j-1]) || 
                        (dp[j-1] && s2[j-1] === s3[i+j-1]);
            }
        }
    }
    
    return dp[s2.length];
}